### PR TITLE
Implement new `hash_memory` proc

### PR DESCRIPTION
**Pull request by Fumuran**
This is the same PR as #1369 needed to re-add this procedure to the `al-migrate-new-padding-rule` branch.

### CODE REVIEW

File: stdlib/tests/crypto/rpo.rs
Comment (bobbinth) [inline stdlib/tests/crypto/rpo.rs:1]: Unless they are already here, could we add tests somewhere here for hashing empty input (i.e., 0 words or 0 elements)? With the new padding rule, hashing empty input should return `[ZERO; 4]`.

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you! I didn't review the logic of the function in detail as I'm assuming it is the same as in the previous PR. But I did leave a couple of small comments inline.

File: stdlib/tests/crypto/rpo.rs [hunk 1]
```diff
@@ -274,3 +274,96 @@ fn test_squeeze_digest() {
...
+        9, 10, 11, 12, 
+        13, 14, 15
+    ]).into_iter().map(|e| e.as_int()).collect();
+    // make sure that value `11` stays unchanged
+    expected_hash.push(11);
+    build_test!(compute_inputs_hash_15, &[]).expect_stack(&expected_hash);
+}
+
+#[test]
+fn test_hash_memory_empty() {
```
Comment (bobbinth) [inline stdlib/tests/crypto/rpo.rs:356]: Let's change this test to also test `hash_memory_words` and `absorb_double_words_from_memory`.

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 2]
```diff
@@ -32,12 +37,14 @@ end
 #! This requires that `end_addr=start_addr + 2n + 1`, otherwise the procedure will enter an infinite
 #! loop. `end_addr` is not inclusive.
```
Comment (bobbinth) [inline stdlib/asm/crypto/hashes/rpo.masm:38]: I'm not sure this is exactly correct. I think proving `end_addr = start_addr` also works fine.

**Review by bobbinth — COMMENTED** [review]
Thank you! I left a couple more comments inline. Also, it may be good to add a `hash_memory_double_words` procedure - but probably in a different PR.

File: stdlib/tests/crypto/rpo.rs (hunk 1 above)
Comment (bobbinth) [reply stdlib/tests/crypto/rpo.rs:335]: Yes, I think we should. Is there a downside of doing this?

**Review by Al-Kindi-0 — APPROVED** [review]
Looks good to me, thank you!

File: stdlib/tests/crypto/rpo.rs (hunk 1 above)
Comment (Fumuran) [reply stdlib/tests/crypto/rpo.rs:335]: No, at least as far as I can see — no, so I lifted this restriction.

File: stdlib/tests/crypto/rpo.rs [hunk 3]
```diff
@@ -354,16 +333,50 @@ fn test_hash_memory() {
...
+
+    begin
+        push.1000 # end address
+        push.1000 # start address
+
+        exec.rpo::hash_memory_words
+    end
+    ";
+
+    build_test!(source, &[]).expect_stack(&[0; 4]);
```
Comment (bobbinth) [inline stdlib/tests/crypto/rpo.rs:367]: Why not test against expected stack `[0; 16]`?

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left one question inline - but it could also be addressed separately.

File: stdlib/tests/crypto/rpo.rs (hunk 3 above)
Comment (bobbinth) [reply stdlib/tests/crypto/rpo.rs:367]: Let's check the entire stack - this way we'll be sure that the procedure doesn't "mess" with the rest of the stack.
